### ISSUE TITLE
Open Docker Desktop at startup without logging on Windows

### ISSUE THREAD

**Issue by AxelGuignard** [docker/roadmap#515]
**Tell us about your request**
It should be possible to start Docker Desktop on windows at startup without having to log in first.

**Which service(s) is this request for?**
Docker for Windows

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**
After a long and frustrating search, I came to the conclusion that it is currently impossible to start Docker Desktop or the docker engine at startup on windows without logging with a user.

Solutions I found on internet were either to configure a task to [launch Docker Desktop at startup](https://github.com/docker/for-win/issues/13096) or to [use the daemon as a service that automatically start](https://www.coretechnologies.com/products/AlwaysUp/Apps/StartDockerDaemonAsAWindowsService.html). None of these seem to work anymore, because now Docker Desktop must be run in interactive mode and it seems that starting dockerd alone don't work with windows, or at least is not supposed to be used that way.

I understand there is a security issue of being able to run docker without logging first, but it makes it impossible to run applications in production reliably using docker, as a server reboot will make them inaccessible until someone actively logs into the server and start Docker Desktop.
This can lead to frequent application outages, that are easily fixed (providing there is someone available and authorized to log into the server and start Docker Desktop), but hinders productivity and customer satisfaction.
There are other cases when this is inconvenient, such as using docker to do CI.

We should be able to automatically start Docker Desktop at server startup in production.

**Are you currently working around the issue?**
As far as I know from my research, the only workaround is to automatically logs a user at startup, but the security implications are way worse so we didn't do that.

**Additional context**
Docker Desktop v4.21.1
Windows Server 2022

**Comment by macmirchdocker** [comment]
What you describe should be possible through the Docker Engine - the instructions for install of this are [available here](https://docs.docker.com/engine/install/binaries/).  This would not be possible with Docker Desktop.

**Comment by AxelGuignard** [comment]
Hi @macmirchdocker.
Since Docker Desktop is the preferred way of installing docker on windows (according to docker documentation), this feature should be available for Docker Desktop.
If windows limitations don't allow that, using the engine directly could indeed be a workaround, except that I believe it is not currently possible to launch the docker engine at windows startup.

I tried installing docker directly in WSL, but once again you [can't start it at windows startup without logging in with a user](https://github.com/microsoft/WSL/issues/9231).

At the moment, there is no way of doing that except with some whacky workaround like auto login.

I don't consider this issue closed.

**Comment by macmirchdocker** [comment]
Re-opening.

**Comment by WhimsicalWalrus** [comment]
Has this issues been revisited at all?

**Comment by fricpa** [comment]
What a pain. I am trying to set up a Windows Server to be a Jenkins Agent but there is no way to enable it to run docker commands because of this. A selection of errors I get when trying various workarounds are:

```
docker: error during connect: in the default daemon configuration on Windows, the docker client must be run with elevated privileges to connect: Post "http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/containers/create": open //./pipe/docker_engine: Das System kann die angegebene Datei nicht finden.
See 'docker run --help'.

C:\Users\Administrator>docker run hello-world
    docker: request returned Internal Server Error for API route and version http://%2F%2F.%2Fpipe%2Fdocker_engine/v1.24/containers/create, check if the server supports the requested API version.
    See 'docker run --help'.

docker: Error response from daemon: hcs::CreateComputeSystem 49df84e247a5146d5a480bf71da3d654c1f815923e13a6febe9e8ccf86c24b6d: Die Anforderung wird nicht unterstützt.
```

**Comment by andyg2** [comment]
Does ARSO help in your situation?
Admittedly it's a variation of automatic login but it also locks the session.

https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/manage/component-updates/winlogon-automatic-restart-sign-on--arso-

**Comment by fricpa** [comment]
I ended up creating another local admin user (besides the default Administrator on Windows Server) and could configure automatic login for that via netplwiz and by following https://www.makeuseof.com/windows-11-missing-auto-login-fix/ to restore the auto login option there. Now I just have a shellscript in shell:startup that definitely runs in user mode and Docker Desktop is started properly and then I can finally run some docker commands in my Jenkins Agent jobs...

**Comment by grazy27** [comment]
> Does ARSO help in your situation? Admittedly it's a variation of automatic login but it also locks the session.
> 
> https://learn.microsoft.com/en-us/windows-server/identity/ad-ds/manage/component-updates/winlogon-automatic-restart-sign-on--arso-

ARSO solved my issue as well. 
Interestingly, it doesn't work if last login was made via RDP, after physically logging in to server once it started working

**Comment by dingdongyo** [comment]
I'm using Docker to run Photoprism, per their instructions. Whenever the PC restarts, I can't access photos anymore until I log into the computer again. I've tried adding restart tasks to Task Scheduler and setting up a service using NSSM, nothing is working. Could really use the ability to automatically start Docker at startup with no login.

**Comment by andyg2** [comment]
> automatically start Docker at startup with no login

The WSL2 starts under user space (at login) but Hyper-V does run before login. If there's any chance of getting Docker to run before login you need to use the Hyper-V option. 

Here's how I would attempt that.

Open settings and uncheck `Use the WSL 2 based engine` - click restart docker and restart system
Then register dockerd as a service and ensure it starts with dependancies.

```ps1
dockerd --register-service
Set-Service vmms -StartupType Automatic    # Hyper-V Virtual Machine Management
Set-Service hns -StartupType Automatic     # Host Network Service
Set-Service hvhost -StartupType Automatic  # Hyper-V Host Compute Service
Set-Service docker -StartupType Automatic  # Docker service
```

Then restart system again and don't log in. You might need to give it a minute.

This is untested but it might help get you there.

**Comment by UlrichThiess** [comment]
Docker Desktop on Windows is only for development. Better is to use wsl with Ubuntu and Docker.

**Comment by dingdongyo** [comment]
> > automatically start Docker at startup with no login
> 
> The WSL2 starts under user space (at login) but Hyper-V does run before login. If there's any chance of getting Docker to run before login you need to use the Hyper-V option.
> 
> Here's how I would attempt that.
> 
> Open settings and uncheck `Use the WSL 2 based engine` - click restart docker and restart system Then register dockerd as a service and ensure it starts with dependancies.
> 
> dockerd --register-service
> Set-Service vmms -StartupType Automatic    # Hyper-V Virtual Machine Management
> Set-Service hns -StartupType Automatic     # Host Network Service
> Set-Service hvhost -StartupType Automatic  # Hyper-V Host Compute Service
> Set-Service docker -StartupType Automatic  # Docker service
> Then restart system again and don't log in. You might need to give it a minute.
> 
> This is untested but it might help get you there.

Apologies for the delay, It's been a while since I could revisit this.

I attempted to follow your instructions, but I don't get the option to uncheck 'Use the WSL2 based engine'. It's greyed out and states "(Windows Home can only run the WSL 2 backend)." Thank you for the idea though!

If it's helpful at all, these are the instructions from PhotoPrism for running it using Docker on Windows: https://docs.photoprism.app/getting-started/docker-compose/#__tabbed_1_5

**Comment by fancybear-dev** [comment]
I truly cannot comprehend how this is still an issue after all these years... Docker says that Docker Desktop is for development only + not supported on Windows Server. Microsoft only has official guides for WSL2 or Windows Containers. But who is going to use Windows Containers? More often than not, you cannot use it. Some organisations simply do not have the knowledge to run Linux in enterprise context, so they don't have another choice than using Windows. Given how often Windows NEEDS to update for e.g. security updates - it's vital that it survives reboots. 

Anyway, I added some gray hairs. I managed to have a seemingly reliable fix running, without hurting security by e.g. automatic sign-in using creds on disk (ARSU did not work for me given RDP use only) and what have you not (it uses S4U, Service for User). Wanted to share it given the amount dread others have already went through.

Tested on Windows Server 2025 Datacenter Azure edition in Azure.

- Install Docker Desktop (choose WSL2!) and go through all the steps until you can start containers.
	- Fun fact, Docker Desktop is officially not supported AT ALL on Windows Server. I have not seen or experienced any issues, but don't go barking at Docker when we are CLEARLY running unsupported setups. 
- Use the magic ps1 script
	- It's designed to be idem-potent for quick iterating
	- Does not require any password on-disk, but does not run as SYSTEM...
	- The 2 minute wait is dirty, it's designed to not overload the OS at boot and supersede any other potential dependencies
	- Downside, Docker Desktop UI is not retrievable after this. Don't know why, also don't care - but if you do, beware. To update Docker Desktop, I recommend Docker Desktop CLI (docker desktop update) or winget (winget upgrade --all --accept-package-agreements --accept-source-agreements)

```
# Define the task details
$taskName = "StartDockerDesktopAtBoot"
$taskDescription = "Starts Docker Desktop automatically after boot (with 2-minute delay), regardless of user login."
$wrapperScriptPath = [Environment]::GetFolderPath("Desktop") + "\DockerStartWrapper.ps1"  # Wrapper script on current user's desktop for delay

try {
    # Get current user's full username for the principal
    $userId = [System.Security.Principal.WindowsIdentity]::GetCurrent().Name

    # Check if the task already exists and remove it to avoid conflicts
    $existingTask = Get-ScheduledTask -TaskName $taskName -ErrorAction SilentlyContinue
    if ($existingTask) {
        Unregister-ScheduledTask -TaskName $taskName -Confirm:$false
        Write-Host "Existing task removed to allow recreation."
    }

    # Create a wrapper script for the 2-minute delay (workaround for -Delay parameter not being supported)
    $wrapperContent = @"
# start service...
echo "Waiting for 2 minutes before starting Docker Desktop..."
Start-Sleep -Seconds 120
echo "Starting service com.docker.service... "
start-service -Name com.docker.service
echo "Starting docker desktop... "
start "$Env:Programfiles\Docker\Docker\Docker Desktop.exe"
echo "Done."
"@
    $wrapperContent | Out-File -FilePath $wrapperScriptPath -Encoding UTF8
    Write-Host "Wrapper script created at $wrapperScriptPath for delay handling."

    # Create the action to run the wrapper script via PowerShell
    $action = New-ScheduledTaskAction -Execute "powershell.exe" -Argument "-NoProfile -ExecutionPolicy Bypass -File `"$wrapperScriptPath`""

    # Create the trigger for at startup (without -Delay to avoid errors)
    $trigger = New-ScheduledTaskTrigger -AtStartup

    # Create the principal: Run as current user with highest privileges
    $principal = New-ScheduledTaskPrincipal -UserId $userId -LogonType S4U -RunLevel Highest

    # Create settings: Allow running on battery, no execution time limit, etc.
    $settings = New-ScheduledTaskSettingsSet -AllowStartIfOnBatteries:$true -DontStopIfGoingOnBatteries:$true -ExecutionTimeLimit ([TimeSpan]::Zero) -Hidden:$false -StartWhenAvailable:$true -Compatibility Win8

    # Create the task object
    $task = New-ScheduledTask -Action $action -Trigger $trigger -Principal $principal -Settings $settings -Description $taskDescription

    # Register the task in Task Scheduler (provide password here)
    Register-ScheduledTask -TaskName $taskName -InputObject $task

    # Output confirmation
    Write-Host "Scheduled task '$taskName' created successfully. It will start Docker Desktop at boot (after a 2-minute delay via wrapper) with highest privileges, as user '$userId'."
} catch {
    Write-Host "Error creating task: $($_.Exception.Message)"
}
```

**Comment by ottelo9** [comment]
@fancybear-dev 
Thanks for your script. I run it and now have the new task and the created ps1 on my desktop. After a reboot the service starts. But the containers wont start automatically and I cant start them on my own via compose (Greenbone Community container). I always get this error: 

`unable to get image 'registry.community.greenbone.net/community/data-objects': request returned 500 Internal Server Error for API route and version http://%2F%2F.%2Fpipe%2FdockerDesktopLinuxEngine/v1.51/images/registry.community.greenbone.net/community/data-objects/json, check if the server supports the requested API version`

Only if I start Docker Desktop manually, then everything works.

**Comment by andyg2** [comment]
In case it helps anyone, the only way to reliably start something before windows login is to use Hyper-V

```powershell
Set-VM -Name $VMName -AutomaticStartAction Start -AutomaticStopAction Shutdown
```

**Comment by alelom** [comment]
ARSO is a solution, but I have to pair it with my JetKVM to ensure "manual" log in, rather than RDP. Luckily, I have my JetKVM device always connected to my home server, as it is useful for tasks like e.g. accessing BIOS. I used it also when I had Linux installed as my main Server OS too. 

This isn't ideal, but the downsides and limitations of Linux as my main server OS still made this worth it. After months of experiments I gave up Linux as no distro does everything I needed (including Proxmox, TrueNAS scale, OpenMediaVault, and more classic Ubuntu server and Debian). With this configuration, I have no issue using my GPU/CUDA drivers, many different workloads, containers, VMs, gaming and streaming services. And it was comparably much more easy to set up.

I would just love it if docker desktop would officially support Windows Server and provide a better solution to be better integrated with WSL and Linux containers at startup without login.

**Comment by aash591** [comment]
use this https://github.com/aash591/windows-docker-boot-startup/blob/main/docker-startatboot.ps1

**Comment by alelom** [comment]
I managed to make it work with a simple Powershell script on Windows Server 2025.

I basically adapted the [powershell script suggested here](https://github.com/docker/for-win/issues/2632#issuecomment-473983144), adding Docker Desktop and CLI to it:

```powershell
Write-Host "##### Attempting to start Docker Desktop and CLI`n #####"
$docker_wait = 10
function Restart-Docker {
  net stop docker
  net stop com.docker.service
  taskkill /IM 'dockerd.exe' /F
  taskkill /IM 'Docker for Windows.exe' /F
  net start docker
  net start com.docker.service
  Start-Process "C:\Program Files\Docker\Docker\Docker Desktop.exe"
  Start-Process "C:\Program Files\Docker\Docker\DockerCli.exe"
}

# Docker For Windows.exe might not be immediately ready
$sleep_multiplier = 0
do {
  Restart-Docker
  $sleep_seconds = $docker_wait * ++$sleep_multiplier
  Write-Output "Sleeping for $sleep_seconds seconds"
  Start-Sleep $sleep_seconds
  docker images
} while ($LASTEXITCODE -ne 0)
Write-Host "`n`n##### Terminating. #####"
```

Then, I created a Task Scheduler action that should run with the highest privileges at startup _whether the user is logged in or not_:

<img width="640" height="489" alt="Image" src="https://github.com/user-attachments/assets/cfcf6392-4f14-4333-b56e-6c8962c7499b" />


And the rest of the configurations as you'd expect (see details below):

<details>

In the Action I included both the script above and Docker Desktop directly (likely redundant):
<img width="638" height="486" alt="Image" src="https://github.com/user-attachments/assets/735f4281-dd40-46ba-834c-233bfdb6d2f2" />

<img width="639" height="483" alt="Image" src="https://github.com/user-attachments/assets/c30521f5-0377-4e9b-acb7-b9ab76a822f6" />

<img width="637" height="488" alt="Image" src="https://github.com/user-attachments/assets/d064646f-7717-4feb-94dc-1983fe52cd32" />

</details>

This made Docker run at startup without the need of log-in, and launching my containers as required!

**Comment by Seithx** [comment]
> I managed to make it work with a simple Powershell script on Windows Server 2025.
> 
> I basically adapted the [powershell script suggested here](https://github.com/docker/for-win/issues/2632#issuecomment-473983144), adding Docker Desktop and CLI to it:
> 
> Write-Host "##### Attempting to start Docker Desktop and CLI`n #####"
> $docker_wait = 10
> function Restart-Docker {
>   net stop docker
>   net stop com.docker.service
>   taskkill /IM 'dockerd.exe' /F
>   taskkill /IM 'Docker for Windows.exe' /F
>   net start docker
>   net start com.docker.service
>   Start-Process "C:\Program Files\Docker\Docker\Docker Desktop.exe"
>   Start-Process "C:\Program Files\Docker\Docker\DockerCli.exe"
> }
> 
> # Docker For Windows.exe might not be immediately ready
> $sleep_multiplier = 0
> do {
>   Restart-Docker
>   $sleep_seconds = $docker_wait * ++$sleep_multiplier
>   Write-Output "Sleeping for $sleep_seconds seconds"
>   Start-Sleep $sleep_seconds
>   docker images
> } while ($LASTEXITCODE -ne 0)
> Write-Host "`n`n##### Terminating. #####"
> Then, I created a Task Scheduler action that should run with the highest privileges at startup _whether the user is logged in or not_:
> 
> <img alt="Image" width="640" height="489" src="https://private-user-images.githubusercontent.com/6352844/550140353-cfcf6392-4f14-4333-b56e-6c8962c7499b.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1NTA0MjcsIm5iZiI6MTc3NTU1MDEyNywicGF0aCI6Ii82MzUyODQ0LzU1MDE0MDM1My1jZmNmNjM5Mi00ZjE0LTQzMzMtYjU2ZS02Yzg5NjJjNzQ5OWIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MDdUMDgyMjA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWQwYjE4Y2NhYjQ5MTA0OTQ0OWQzNTMzOThiMTZkOGY3ODk2NDJmMmU5ODQ5NTlkYzg3OTY1ZjBjZmU5Y2VjYSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vC9g9VgmmMEOWBAxmKqlo3b3GGewd6_fCQBhFhf0Miw">
> And the rest of the configurations as you'd expect (see details below):
> 
> Details
> In the Action I included both the script above and Docker Desktop directly (likely redundant): <img alt="Image" width="638" height="486" src="https://private-user-images.githubusercontent.com/6352844/550140424-735f4281-dd40-46ba-834c-233bfdb6d2f2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1NTA0MjcsIm5iZiI6MTc3NTU1MDEyNywicGF0aCI6Ii82MzUyODQ0LzU1MDE0MDQyNC03MzVmNDI4MS1kZDQwLTQ2YmEtODM0Yy0yMzNiZmRiNmQyZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MDdUMDgyMjA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ODk4ZDNlYTcxN2JmM2M5ODM4Y2Q0ZjM5MmU5OTQ0MTdmYjk4ZTY2N2M0Y2Q2OTg2NjhhOWFhMWJlZjQxZjNiMiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.Ow8rgE2Zl3N4S4DMSCRfCie8UWlondTiSz998ee6-1g">
> 
> <img alt="Image" width="639" height="483" src="https://private-user-images.githubusercontent.com/6352844/550140461-c30521f5-0377-4e9b-acb7-b9ab76a822f6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1NTA0MjcsIm5iZiI6MTc3NTU1MDEyNywicGF0aCI6Ii82MzUyODQ0LzU1MDE0MDQ2MS1jMzA1MjFmNS0wMzc3LTRlOWItYWNiNy1iOWFiNzZhODIyZjYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MDdUMDgyMjA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjk1ZTE2NTY4ZWUyNWMxMWVjZDAwYTVlOTc1YTJkZmNjZjk0MGY1ODVhOGJlNzUwOGZjODRkOWQ0ZmI4NWMxZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.W3vNTsX7Xkrrh_pf8e2cv9q896ZMkOFbwZWUH3TW68A"> <img alt="Image" width="637" height="488" src="https://private-user-images.githubusercontent.com/6352844/550140499-d064646f-7717-4feb-94dc-1983fe52cd32.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzU1NTA0MjcsIm5iZiI6MTc3NTU1MDEyNywicGF0aCI6Ii82MzUyODQ0LzU1MDE0MDQ5OS1kMDY0NjQ2Zi03NzE3LTRmZWItOTRkYy0xOTgzZmU1MmNkMzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MDdUMDgyMjA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mjk0OGY0ZDBlYWExNjQzNzE4ZWFkYjdmMDMzYTdmNzdkMzE5ZTM4MjJiNDc5MmE3OTY3NDVkYjRhMGE1YTc5ZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.51qc5K1Pj8s8tG-T7bFNTK1t1pfmJNHf37-TixIWgws">
> This made Docker run at startup without the need of log-in, and launching my containers as required!

Will it work on Win10/Win11 then, you reckon?
